### PR TITLE
switch to http4s (fixes #675)

### LINKED ISSUE #675: switch over to http4s
In the process, we can clean up some things:

 * Always use HTTP `Warning` header for warnings.
 * Always use non-200 HTTP status line for a one-line error summary.
 * Always put detailed error and warning messages in the response in a standardized way (e.g. `{response: a, errors: [], warnings: []}` or some such).
 * Recognize `Accept` and `Content-Type` headers and produce correct `Content-Type` headers.

### PR DESCRIPTION
**Pull request by mossprescott**
The services (/metadata, /data, /query, and /compile) are now broken out
as separate HttpService instances, with path prefixes used to map each one
under the right path. This will make it easier to add additional APIs in
the future.

That meant making a couple of very small changes to the tests so that we
hit OPTIONS under a valid path.

One other test change was to make the test for a path with encoded spaces in it
actually check the result.

Other than that, this is just mapping the unfiltered API to http4s's. A couple
of things were missing and required workarounds:

1. `unfiltered.util.Port.any`: finds a "random", unbound port for testing purposes. For now, just using 8888.
2. URL decoding of path segments. Will possibly submit a PR to http4s for this.
3. Handling of paths with trailing slashes. These are swallowed by http4s's Path extractors. For now, just treating paths as directory paths when needed. Submitted a bug to http4s, may need to revisit when it's fixed.
4. Had to re-implement unfiltered's "wait for keystroke" behavior. It's ugly, but seems to work well enough.

http4s pulled in scalaz-stream 0.6a, so I updated our dependency. There was only
one minor code change required to fix a deprecation warning.

### CODE REVIEW

**Comment by jdegoes** [comment]
@mossprescott README.md should be updated too, right?

**Comment by jdegoes** [comment]
@sellout You had a chance to review yet?

File: web/src/main/scala/slamdata/engine/api/fs.scala [hunk 1]
```diff
@@ -24,191 +25,216 @@ import scalaz.stream._
...
+  private def dataSourceFor(path: Path): Task[Response] \/ (FileSystem, Path) =
+    lookupBackend(path).map { case (backend, _, relPath) => (backend.dataSource, relPath) }
+
+  private def errorResponse(status: org.http4s.dsl.impl.EntityResponseGenerator, e: Throwable) = {
+    e match {
+      case PhaseError(phases, causedBy) => status(Json.obj(
+          "error"  := causedBy.getMessage,
+          "phases" := phases))
+
+      case _ => status(e.getMessage)
```
Comment (jdegoes) [inline web/src/main/scala/slamdata/engine/api/fs.scala]: Hopefully http4s will trim this to reasonable length if necessary as I believe our errors can sometimes get gigantic.

File: web/src/main/scala/slamdata/engine/api/fs.scala [hunk 2]
```diff
@@ -24,191 +25,216 @@ import scalaz.stream._
...
+        case (pe @ PathError(_)) :: Nil => errorBody(BadRequest, pe :: Nil)
+        case es                         => errorBody(InternalServerError, es)
+      }
+    } yield Ok("")).fold(identity, identity)
+  }
 
-    case _ => ResponseString(e.getMessage)
+  object AsPath {
+    def unapply(p: HPath): Option[Path] = {
+      Some(Path("/" + p.toList.map(java.net.URLDecoder.decode(_, "UTF-8")).mkString("/")))
```
Comment (jdegoes) [inline web/src/main/scala/slamdata/engine/api/fs.scala]: I think decode may throw exceptions and / or return null in various cases, so this code isn't safe.

File: scripts/testJar [hunk 3]
```diff
@@ -31,17 +31,17 @@ cat $SLAMDATA_LOG_FILE
 
 echo "Killing SlamData JAR..."
 
-kill -9 $SLAM_PID
-KILL_STATUS=$?
+kill -9 $SLAM_PID || echo '... not killed'
+# KILL_STATUS=$?
```
Comment (jdegoes) [inline scripts/testJar]: Dead code?

File: scripts/testJar [hunk 4]
```diff
@@ -31,17 +31,17 @@ cat $SLAMDATA_LOG_FILE
...
 
-echo "Status code from killing SlamData: $KILL_STATUS"
+# echo "Status code from killing SlamData: $KILL_STATUS"
```
Comment (jdegoes) [inline scripts/testJar]: Debug code?

File: web/src/main/scala/slamdata/engine/api/package.scala [hunk 5]
```diff
@@ -3,40 +3,24 @@ package slamdata.engine
...
+  // Note: CORS middleware is comming in http4s post-0.6.5
+  def `Access-Control-Allow-Origin`(value: String) = Header("Access-Control-Allow-Origin", value)
+  val AccessControlAllowOriginAll = `Access-Control-Allow-Origin`("*")
+
+  def `Access-Control-Allow-Methods`(methods: List[String]) = Header("Access-Control-Allow-Methods", methods.mkString(", "))
+  def `Access-Control-Allow-Headers`(headers: List[HeaderKey]) = Header("Access-Control-Allow-Headers", headers.map(_.name).mkString(", "))
+  def `Access-Control-Max-Age`(seconds: Long) = Header("Access-Control-Max-Age", seconds.toString)
+
+  object Destination extends HeaderKey.Singleton {
+    type HeaderT = Header // FIXME: a big hack?
```
Comment (jdegoes) [inline web/src/main/scala/slamdata/engine/api/package.scala]: Why is this necessary?

File: web/src/main/scala/slamdata/engine/api/server.scala [hunk 6]
```diff
@@ -10,19 +7,34 @@ import slamdata.engine.config._
...
+      .mountService(api.queryService,    "/query/fs")
+      .mountService(api.compileService,  "/compile/fs")
+      .mountService(api.metadataService, "/metadata/fs")
+      .mountService(api.dataService,     "/data/fs")
+      .start
   }
 
+  def runAndWaitForInput(port: Int, fs: FSTable[Backend]): Task[Unit] = for {
+    server <- run(port, fs)
+    _ = println("Embedded server listening at port " + port)
```
Comment (jdegoes) [inline web/src/main/scala/slamdata/engine/api/server.scala]: Effectful code!!! Think you want: `_ <- Task.delay(...)` for all these lines which are currently `_ = `.

File: scripts/testJar (hunk 3 above)
Comment (mossprescott) [inline scripts/testJar]: At the moment, yes. The idea was that the server should start and then need to be killed, as it was before, but somehow when this script runs the server it immediately exits. I couldn't figure out what was going on, so I just punted and made it not fail.

I can clean up this dead code, but the more important question is whether we're concerned about the server quitting at all.

File: web/src/main/scala/slamdata/engine/api/fs.scala (hunk 1 above)
Comment (mossprescott) [inline web/src/main/scala/slamdata/engine/api/fs.scala]: I'm not sure. I believe this can only be hit if we have a bug, because any legitimate failure should be captured as a PhaseError.

File: web/src/main/scala/slamdata/engine/api/fs.scala (hunk 2 above)
Comment (mossprescott) [inline web/src/main/scala/slamdata/engine/api/fs.scala]: I don't *think* so, based on the documentation, which says it throws only if the "named character encoding is not supported", and also explicitly says using anything other than UTF-8 "may introduce incompatibilities". I can't think of a case that *should* provoke it to return null, but I have no way of being sure.

What do we need to feel comfortable using a Java class as if it were pure? Obviously I could wrap with catch and Option to be safe.

Comment (jdegoes) [inline web/src/main/scala/slamdata/engine/api/fs.scala]: If the docs guarantee it won't return null or throw an exception, then I think it's OK to treat it as pure.

File: web/src/main/scala/slamdata/engine/api/package.scala (hunk 5 above)
Comment (mossprescott) [inline web/src/main/scala/slamdata/engine/api/package.scala]: I can't define the `Header` the way http4s does the standard headers, because they made some of their implementation types private, and this is what I could figure out that worked. I suppose it leaves us with less type safety than the other Headers provide, but I'm not sure in exactly what way. 

I'll remove the FIXME, at least.

File: web/src/main/scala/slamdata/engine/api/server.scala (hunk 6 above)
Comment (mossprescott) [inline web/src/main/scala/slamdata/engine/api/server.scala]: Doh! This was refactored out of `main`, which of course gets a pass. I'll fix it.

File: web/src/test/scala/slamdata/engine/api/fs.scala [hunk 7]
```diff
@@ -23,33 +21,22 @@ import com.ning.http.client.{Response}
 
 class ApiSpecs extends Specification with DisjunctionMatchers with PendingWithAccurateCoverage {
   sequential  // Each test binds an arbitrary port
```
Comment (sellout) [inline web/src/test/scala/slamdata/engine/api/fs.scala]: This comment is no longer true, is it?

**Comment by mossprescott** [comment]
Mostly minor adjustments, but it turns out testJar was catching the server actually quitting in a way that we almost certainly don't want, so that is fixed.
